File: src/telemetry/wait_ready.cpp
#include "telemetry/wait_ready.h"

#include <google/protobuf/util/json_util.h>
#include <tuple>

namespace Serializers
{
namespace Telemetry
{
ReplayReady::ReplayReady(const PbTelemetry::ReplayReady& protobuf) {
    status = static_cast<ReadyStatus>(protobuf.status());
}

ReplayReady::operator PbTelemetry::ReplayReady() const {
    PbTelemetry::ReplayReady ret;
    ret.set_status(static_cast<PbTelemetry::ReadyStatus>(status));
    return ret;
}

std::string ReplayReady::serializeAsJsonString() const {
    PbTelemetry::ReplayReady protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string ReplayReady::serializeAsProtobufString() const {
    PbTelemetry::ReplayReady protobuf(*this);
    return protobuf.SerializeAsString();
}

bool ReplayReady::deserializeFromJsonString(const std::string& str) {
    PbTelemetry::ReplayReady protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool ReplayReady::deserializeFromProtobufString(const std::string& str) {
    PbTelemetry::ReplayReady protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

ReplayStart::ReplayStart(const PbTelemetry::ReplayStart& protobuf) {
    startTimestamp = protobuf.starttimestamp();
}

ReplayStart::operator PbTelemetry::ReplayStart() const {
    PbTelemetry::ReplayStart ret;
    ret.set_starttimestamp(startTimestamp);
    return ret;
}

std::string ReplayStart::serializeAsJsonString() const {
    PbTelemetry::ReplayStart protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string ReplayStart::serializeAsProtobufString() const {
    PbTelemetry::ReplayStart protobuf(*this);
    return protobuf.SerializeAsString();
}

bool ReplayStart::deserializeFromJsonString(const std::string& str) {
    PbTelemetry::ReplayStart protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool ReplayStart::deserializeFromProtobufString(const std::string& str) {
    PbTelemetry::ReplayStart protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}
}
}

File: py/telemetry/wait_ready.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict

from telemetry import wait_ready_pb2
from google.protobuf.json_format import MessageToJson, Parse

class ReadyStatus(Enum):
    WAITING = 0
    READY = 1


@dataclass
class ReplayReady:
    status: ReadyStatus = None
    
    _proto_message: wait_ready_pb2.ReplayReady = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = wait_ready_pb2.ReplayReady()

    def _populate_proto(self):
        self._proto_message.status = self.status.value

    @classmethod
    def _from_proto(cls, proto_message) -> "ReplayReady":
        return cls(
            status = ReadyStatus(proto_message.status),
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "ReplayReady":
        message = wait_ready_pb2.ReplayReady()
        message.ParseFromString(data)
        return cls(
            status = ReadyStatus(message.status),
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "ReplayReady":
        message = wait_ready_pb2.ReplayReady()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class ReplayStart:
    startTimestamp: int = 0
    
    _proto_message: wait_ready_pb2.ReplayStart = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = wait_ready_pb2.ReplayStart()

    def _populate_proto(self):
        self._proto_message.startTimestamp = self.startTimestamp

    @classmethod
    def _from_proto(cls, proto_message) -> "ReplayStart":
        return cls(
            startTimestamp = proto_message.startTimestamp,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "ReplayStart":
        message = wait_ready_pb2.ReplayStart()
        message.ParseFromString(data)
        return cls(
            startTimestamp = message.startTimestamp,
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "ReplayStart":
        message = wait_ready_pb2.ReplayStart()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())
